BlueDBOnDisk/src/main/java/io/bluedb/disk/recovery/RecoveryManager.java

```diff
@@ -96,20 +96,10 @@ protected void cleanupHistory() throws BlueDbException {
 		}
 		List<TimeStampedFile> timestampedFiles = Blutils.map(files, (f) -> new TimeStampedFile(f) );
 		Collections.sort(timestampedFiles);
-
-		int firstChangeToKeep = 0;
-		int lastChangeToKeep = -1;
-		
-		for (int i=0; i<files.size(); i++) {
-			TimeStampedFile timeStampedFile = timestampedFiles.get(i);
-			if (timeStampedFile.getTimestamp() < backupStartTime) {
-				firstChangeToKeep = i;
-			}
-			if (timeStampedFile.getTimestamp() <= backupEndTime) {
-				lastChangeToKeep = i;
-			}
-		}
-		
+		// Note: the last change before backup might by incomplete so we'll include it
+		int lastChangeBeforeBackup = Blutils.lastIndex(timestampedFiles, (t) -> t.getTimestamp() < backupStartTime );
+		int firstChangeToKeep = Math.max(0,  lastChangeBeforeBackup);
+		int lastChangeToKeep = Blutils.lastIndex(timestampedFiles, (t) -> t.getTimestamp() <= backupEndTime );

```

Comment:
You could get the lastChangeToKeep first then pass in the index to start at for lastChangeBeforeBackup to avoid wasting checking items that you know are outside the time you're looking for. There probably isn't ever going to be enough items to really matter though right?

Reply:
I don't think efficiency matters much here.  It's only going to iterate over maybe a few thousand every time you backup.